app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -133,16 +125,31 @@ def delete_object(ar_object, hierarchy)
     # FYI if there is no such association (e.g. a method name), nil will be returned (counts as not by association).
     # Another possible confusion with hand-crafted hierarchies may occur when before a
     # "Plain-..." entry, there is set an unrelated "Association-..." entry, then incorrect association will be set.
-    match = hierarchy.last&.match(ASSOCIATION_RE)
-    association = match[:klass].constantize.reflect_on_association(match[:association]) if match
+    if ar_object.destroyed_by_association
+      reflection = ar_object.destroyed_by_association
+    else
+      # the association was not cached
+      match = hierarchy.last&.match(ASSOCIATION_RE)
+      reflection = match[:klass].constantize.reflect_on_association(match[:association]) if match
+    end
     # doing without a reload is a mess related to Service, Proxy and PaymentGatewaySettings, maybe improve next life
     ar_object.reload unless ar_object.class.background_deletion.empty?
-    ar_object.destroyed_by_association = association
-    ar_object.background_deletion_method_call
+    ar_object.destroyed_by_association = reflection
+    call_delete_method(ar_object)
 
     evict_object(ar_object)
   end
 
+  def call_delete_method(ar_object)
+    association = ar_object.destroyed_by_association
+    method = association&.options&.fetch(:dependent, nil) == :delete ? :delete : :destroy!
+    ar_object.send(method)
+  end
```

Comment:
So the delete method is taken from the association, but what happens with the root object of the hierarchy? is it always deleted by `destroy!`?

Reply:
thanks for clarifying!